src/main/java/net/dries007/tfc/common/blocks/soil/IMudBlock.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed under the EUPL, Version 1.2.
+ * You may obtain a copy of the Licence at:
+ * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
+ */
+
+package net.dries007.tfc.common.blocks.soil;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.InteractionHand;
+import net.minecraft.world.InteractionResult;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.item.alchemy.PotionUtils;
+import net.minecraft.world.item.alchemy.Potions;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.material.Fluids;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+
+import net.dries007.tfc.common.capabilities.Capabilities;
+import net.dries007.tfc.common.fluids.FluidHelpers;
+import net.dries007.tfc.config.TFCConfig;
+import net.dries007.tfc.util.Helpers;
+
+/**
+ * Blocks that can be turned into their mud forms
+ */
+public interface IMudBlock
+{
+    /**
+     * Gets the mud block this block will transform into when interacted with
+     */
+    BlockState getMud();
+
+    /**
+     * Transforms this block into mud if the player has the required amount of water in their container
+     */
+    default InteractionResult transformToMud(BlockState mud, Level level, BlockPos pos, Player player, InteractionHand hand)
+    {
+        if (!TFCConfig.SERVER.enableDirtToMudCreation.get())
+        {
+            return InteractionResult.PASS;
+        }
+
+        ItemStack stack = player.getItemInHand(hand);
+        final int waterRequired = 100;
+        final FluidStack water = new FluidStack(Fluids.WATER, waterRequired);
+
+        if (stack.getItem().equals(Items.WATER_BUCKET))
+        {
+            // Unable to deplete fluid in bucket, ignore interaction
+            return InteractionResult.PASS;
+        }
+
+        if (PotionUtils.getPotion(stack) == Potions.WATER)
+        {
+            if (!player.isCreative())
+            {
+                player.setItemInHand(hand, new ItemStack(Items.GLASS_BOTTLE));
+            }
+
+            setMud(mud, level, pos, water);
+            return InteractionResult.SUCCESS;
+        }
+
+        IFluidHandler fluidHandler = Helpers.getCapability(stack, Capabilities.FLUID_ITEM);
+
+        if (fluidHandler != null && fluidHandler.getFluidInTank(0).containsFluid(water))
+        {
+            fluidHandler.drain(waterRequired, player.isCreative() ? IFluidHandler.FluidAction.SIMULATE : IFluidHandler.FluidAction.EXECUTE);
```

Comment:
No. This is not how to interact with a fluid container.

- Simulate the drain first, then drain
- Don't query specific tanks directly, nor use `containsFluid` on specific tanks
- Only execute what you have a priori simulated

Look at `FluidHelpers.transferBetweenItemAndOther` for handling (all) buckets, not just hardcoded water ones, and handling arbitrary fluid containers properly.

https://github.com/TerraFirmaCraft/TerraFirmaCraft/blob/52642eba614929df5a12ea5aaa69293fc510b8dd/src/main/java/net/dries007/tfc/common/fluids/FluidHelpers.java#L207

Reply:
I hope my new implementation addresses these issues? It's not entirely like `FluidHelpers.transferBetweenItemAndOther`, but works well in my testing. It still prevents buckets from converting dirt to mud as I feel that might be annoying if trying to place water sources on dirt.